src/Typechecker/Typechecker.hs

```diff
@@ -105,23 +105,15 @@ exprtype (Ref s) = do
 exprtype (Tuple xs) = do
   xs' <- mapM exprtype xs
   return $ Tup xs'
-exprtype (App n es) = do -- FIXME. Tuple composition is bad.
-  es' <- exprtype es
-  let es'' = case es' of
-        Tup [Tup xs] -> Tup xs
-        x -> x
+exprtype (App n es) = do
+  est <- exprtype es
   _t <- getType n
   case _t of
     (Function (Ft i o)) -> do
-      if es'' == i then
-        return o -- supplied param matches expected input type
-      else
-        do
-          -- otherwise, unify the supplied param w/ the input type
-          u <- unify (es'') i -- oof
-          case u of
-            -- verify the unified result is ultimately the same as the input type
-            x1 -> if x1 == i then return o else mismatch (Plain es'') (Plain i)
```

Comment:
I would be a little bit careful about removing this portion above. This was originally used as part of patch #126, to prevent bad parameter types. I don't think I was able to get a solid test on this in place, so I would take care to keep track that it doesn't regress. If we're in the middle of making changes for the TC that's okay, just want to make sure we don't lose track of potential bugs.

Reply:
I am certain about the change, but I should have added a test. I will add some to the illTyped directory.